mcs/class/Mono.Posix/Mono.Unix/UnixMarshal.cs

```diff
@@ -325,27 +325,34 @@ public static IntPtr StringToHeap (string s, int index, int count, Encoding enco
 			if (encoding == null)
 				throw new ArgumentNullException ("encoding");
 
-			int min_byte_count = encoding.GetMaxByteCount(1);
-			char[] copy = s.ToCharArray (index, count);
-			byte[] marshal = new byte [encoding.GetByteCount (copy) + min_byte_count];
+			int null_terminator_count = encoding.GetMaxByteCount (1);
+			int lengthWithoutNull = encoding.GetByteCount(s);
+			int marshalLength = lengthWithoutNull + null_terminator_count;
```

Comment:
Whatever happened to my comment to use `checked()` here?

You should use `checked()` here. 64-bit mono *can* create strings with length `int.MaxValue` -- if unreliably -- which could cause this to overflow:

```csharp
var s = new string ('a', int.MaxValue); // lots of 'a's
var p = UnixMarshal.StringToHeap (s, 0, int.MaxValue); // exploitable?
```